src/Microsoft.ComponentDetection.Detectors/pip/PipCommandService.cs

```diff
@@ -0,0 +1,123 @@
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+public class PipCommandService : IPipCommandService
+{
+    private readonly ICommandLineInvocationService commandLineInvocationService;
+    private readonly IPathUtilityService pathUtilityService;
+    private readonly IFileUtilityService fileUtilityService;
+    private readonly IEnvironmentVariableService environmentService;
+    private readonly ILogger<PipCommandService> logger;
+
+    public PipCommandService()
+    {
+    }
+
+    public PipCommandService(
+        ICommandLineInvocationService commandLineInvocationService,
+        IPathUtilityService pathUtilityService,
+        IFileUtilityService fileUtilityService,
+        IEnvironmentVariableService environmentService,
+        ILogger<PipCommandService> logger)
+    {
+        this.commandLineInvocationService = commandLineInvocationService;
+        this.pathUtilityService = pathUtilityService;
+        this.fileUtilityService = fileUtilityService;
+        this.environmentService = environmentService;
+        this.logger = logger;
+    }
+
+    public async Task<bool> PipExistsAsync(string pipPath = null)
+    {
+        return !string.IsNullOrEmpty(await this.ResolvePipAsync(pipPath));
+    }
+
+    private async Task<string> ResolvePipAsync(string pipPath = null)
+    {
+        var pipCommand = string.IsNullOrEmpty(pipPath) ? "pip" : pipPath;
+
+        if (await this.CanCommandBeLocatedAsync(pipCommand))
+        {
+            return pipCommand;
+        }
+
+        return null;
+    }
+
+    private async Task<bool> CanCommandBeLocatedAsync(string pipPath)
+    {
+        return await this.commandLineInvocationService.CanCommandBeLocatedAsync(pipPath, new List<string> { "pip3" }, "--version");
+    }
+
+    public async Task<(PipInstallationReport Report, FileInfo ReportFile)> GenerateInstallationReportAsync(string path, string pipExePath = null)
+    {
+        if (string.IsNullOrEmpty(path))
+        {
+            return (new PipInstallationReport(), null);
+        }
+
+        var pipExecutable = await this.ResolvePipAsync(pipExePath);
+        var formattedPath = this.pathUtilityService.NormalizePath(path);
+        var workingDir = new DirectoryInfo(this.pathUtilityService.GetParentDirectory(formattedPath));
+
+        CommandLineExecutionResult command;
+        var reportName = Path.GetRandomFileName();
+        var reportFile = new FileInfo(Path.Combine(workingDir.FullName, reportName));
+
+        string pipReportCommand;
+        if (path.EndsWith(".py"))
+        {
+            pipReportCommand = $"install -e .";
+        }
+        else if (path.EndsWith(".txt"))
+        {
+            pipReportCommand = "install -r requirements.txt";
+        }
+        else
+        {
+            // Failure case, but this shouldn't be hit since detection is only running
+            // on setup.py and requirements.txt files.
+            return (new PipInstallationReport(), null);
```

Comment:
Log this?

Reply:
I'm not concerned on this one - the detector file handling logic should avoid this case.